Resizable side bar

goto-bus-stop opened the pull request:
Demo: https://upload.clinicjs.org/public/757868190b709ae78df7ba4f032acb8fc309aa87695f66bf2cfc4b22c0b55608/2379.clinic-bubbleprof.html

This makes the vertical blue line draggable. Not that much to it! The goal is to make the area chart more usable.

The area chart scales with the width of the sidebar, maintaining aspect ratio.

I'm not super happy with this code yet, so I'll keep tweaking a bit. There are some magic numbers in here that seem unnecessary. eg. when dragging, I add +24 to the X coord to account for a `margin-left`. We also need some visual indicator that the vertical line is draggable, I think. Right now you can only discover it by hovering over it and seeing the cursor change.

Question(s):

- After dragging, this calls `topMostUI.redrawLayout()`, which I took from the window resize event handler. However, it doesn't really seem to do what I expect it would do. It's most clear when first making the sidebar very large, then clicking into a node, then making the sidebar very small. Instead of recalculating the layout, it seems to just make the svg larger and scale everything up. Maybe I'm missing something?
- The SideBarDrag element handles the area chart resize. I don't think that's a big problem, but I had to hack around a bit to do it:
   ```js
    // FIXME this is probably a private API. Is this alright or should the sidebar be its own HtmlContent subclass so it can handle this internally?
    const callbacksOverTime = sideBar.content.get('area-chart')
   ```
   as mentioned in that code comment, I don't think I'm supposed to access `.content` in that way. Nothing else does it. Maybe there's a better way? Or we could have a `.getContentById()` method that just does the above, but more "formally"?

mcollina commented:
Can you add a toggle as well? Ideally something visible/hidden?

Also, the following is a bug:

<img width="487" alt="schermata 2018-09-07 alle 10 46 02" src="https://user-images.githubusercontent.com/52195/45208641-6042f800-b28b-11e8-9d7d-4c729309271f.png">

The call stack and sidebar should move together (or maybe just hide "call stack" if we do not have one)

I think we should redraw only once the bar is settled (maybe with a timeout?)

AlanSl commented:
Regarding toggle - I'll add toggle close / open, minimise / maximise buttons in a separate PR building on this one, I want to try some things re design / button placement

goto-bus-stop commented:
> The call stack and sidebar should move together (or maybe just hide "call stack" if we do not have one)

I thought that'd make the stack frame overlay unusable when the sidebar is wider, but it's not too bad:
![image](https://user-images.githubusercontent.com/1006268/45209288-ddbb3800-b28c-11e8-8bd9-dc5b7aac90bb.png)
probably can do that. if you make the node link graph any smaller than this though it becomes unusable. Might just prevent dragging any further than this

> I think we should redraw only once the bar is settled (maybe with a timeout?)

the node graph is only redrawn once the bar is settled, the sidebar is redrawn all the time though. Initially I had it only redraw everything once the bar is settled but performance seemed good enough to redraw the sidebar all the time (and it's a bit easier). I'll upload a sample that only redraws when dragging stops in a bit so we can compare

AlanSl commented:
Regarding the redraw / debounce question - I've had a look and here's what I recommend:

 - Debounce the call to redrawLayout similar to `onWindowResizeBegin` and `onWindowResizeEnd` (moving and reusing that code if possible to avoid duplication). `redrawLayout` is potentially expensive on complex views because it involves a lot of recalculations, and the elements respond to it assuming that the layout data has changed, so there might be some unnecessary junking and recreating of elements
 - Call `this.topMostUI.draw()` without debouncing - this redraws based on existing data which is enough to get the area chart axis to respace itself nicely etc
 - I've got some work in progress improving the performance of the area chart code more generally; in particular, there's something I can probably split out into a mini-PR moving `createPathsForLayout()` out from `draw()` - when this is done, `AreaChart.draw()` will be  very fast and fine to run every time

AlanSl commented:
Here's that related PR mentioned above: https://github.com/nearform/node-clinic-bubbleprof/pull/265

goto-bus-stop commented:
that sample I promised: https://upload.clinicjs.org/public/4a407c9df7245a6daf3b055ee575b4bb2f519009ef622f30a6a277e8441722d5/2379.clinic-bubbleprof.html Here it only commits the width changes once, at the end of the drag operation. (few ui bugs because I switched to the constantly updating version at that point)

goto-bus-stop commented:
I don't think that debouncing `redrawLayout()` here is necessary since it only happens at the end of the drag operation, not while you're still dragging. If you peek through the "redrawing" overlay at high screen brightness you can see the SVG scaling linearly, not adjusting the shape in any way. It's extremely unlikely that it'll be called more than once per second (you'd have to let go and start dragging again very quickly)

AlanSl commented:
Ah right, looking closer it already has the behaviour I was asking for, and the juddering I was seeing in some profiles was because of that issue fixed in the other PR. After merging in master, resizing seems smooth everywhere except the crazy cluster on 1005, which has other reasons for poor performance. Here's a large profile:

https://upload.clinicjs.org/public/d500208f8f5aeed1906c92764f997a363442170414eb36df70a625c8308ea9b9/7078.clinic-bubbleprof.html

One small bug: for some reason if you drag the sidebar far enough that the node link svg would be less than 300px wide, its width stays at 300px (browser default width?) and it overlaps the sidebar, preventing it from being dragged back again:

![image](https://user-images.githubusercontent.com/29628323/45229528-ebd57c80-b2bd-11e8-8d23-b9ad96f95a9e.png)

Setting `svg.bubbleprof { min-width: 0; }` seems to sort it out.

goto-bus-stop commented:
Ah, that didn't happen in Firefox so I hadn't noticed :sweat_smile: Added the min-width, and also changed the size limits a bit so the graph is at least somewhat usable even at its narrowest size (400px instead of 250px).

AlanSl commented:
Good to hear you by default test in Firefox, I use Chrome by default so we'll catch bugs more quickly :-)

Two bugs unfortunately... (this is testing in Chrome):

1: Often if a sublayout is open and you resize, the loading animation and overlay never disappear. Not sure what might be causing this.

![image](https://user-images.githubusercontent.com/29628323/45436632-2a4aad00-b6ab-11e8-82de-656991eb9f4b.png)

2: If you resize the sidebar, then resize the window down to portrait view, the applied sizes from the resizing break the portrait view. 

![image](https://user-images.githubusercontent.com/29628323/45437010-f9b74300-b6ab-11e8-90ce-f20330f993f6.png)

Two approaches I can see to fixing this last one: 
 a. detect change from landscape to portrait and reset resize (I don't like this so much, don't want too much JS activity on every window resize and my instinct is that it could be glitchy), 
 b. CSS-only: set widths of the resized elements with `!important` in portrait view, therefore overriding the applied styles. This seems like a more robust solution since CSS media queries are what apply the portrait view, and it seems like a reasonable use of `!important` since we're competing with inline styles not CSS - just add a comment next to the `width: X% !important;` s so people know it's there for a good reason (not just a lazy CSS priorities hack)

goto-bus-stop commented:
Agreed re: CSS-only solution. I hadn't seen the loading issue post-drag yet, but just reproduced it accidentally, so will be able to check that out!

goto-bus-stop commented:
Fixed the Redrawing problem—instead of applying the 'redraw' class to the topmost UI element, it now applies it to the original UI always. That's because the original UI is the only one that contains the loading indicator, sublayouts don't.

The buggy behaviour was because a leftover from a previous iteration applied the 'redraw' class to the original UI, but the dragstart and dragend handlers applied it to the topmost UI. If the original UI was in focus, that would work, but if a sublayout was in focus the 'redraw' class was never removed from the original UI.
